### PR TITLE
[WIP] Add metadata option to publish

### PR DESCRIPTION
**Pull request by shubham1172**
Signed-off-by: Shubham Sharma <shubhash@microsoft.com>

# Description

TODO: tests

## Issue reference

Please reference the issue this PR will close: #304 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by shubham1172** [comment]
ping @XavierGeerinck

File: src/implementation/Client/HTTPClient/pubsub.ts [hunk 1]
```diff
@@ -25,9 +27,11 @@ export default class HTTPClientPubSub implements IClientPubSub {
     this.logger = new Logger("HTTPClient", "PubSub", client.getOptions().logger);
   }
 
-  async publish(pubSubName: string, topic: string, data: object = {}): Promise<boolean> {
+  async publish(pubSubName: string, topic: string, data: object = {}, metadata?: KeyValueType): Promise<boolean> {
+    const queryParams = getHTTPMetadataQueryParam(metadata ?? {});
```
Comment (XavierGeerinck) [inline src/implementation/Client/HTTPClient/pubsub.ts:31]: Shouldn't we just pass metadata and make it default in the `getHTTPMetadataQueryParam`?

File: src/implementation/Server/GRPCServer/pubsub.ts [hunk 2]
```diff
@@ -26,8 +27,8 @@ export default class DaprPubSub implements IServerPubSub {
     this.logger = new Logger("GRPCServer", "PubSub", server.client.options.logger);
   }
 
-  async subscribe(pubSubName: string, topic: string, cb: TypeDaprPubSubCallback): Promise<void> {
-    this.logger.info(`Registering onTopicEvent Handler: PubSub = ${pubSubName}; Topic = ${topic}`)
-    this.server.getServerImpl().registerPubSubSubscriptionHandler(pubSubName, topic, cb);
+  async subscribe(pubSubName: string, topic: string, cb: TypeDaprPubSubCallback, _route?: "", metadata?: KeyValueType): Promise<void> {
+    this.logger.info(`Registering onTopicEvent Handler: PubSub = ${pubSubName}; Topic = ${topic}; Metadata: ${metadata}`)
+    this.server.getServerImpl().registerPubSubSubscriptionHandler(pubSubName, topic, cb, metadata ?? {});
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/pubsub.ts:32]: Shouldn't we just pass metadata and make it default in the `registerPubSubSubscriptionHandler`?

File: src/implementation/Server/HTTPServer/pubsub.ts [hunk 3]
```diff
@@ -27,13 +28,13 @@ export default class HTTPServerPubSub implements IServerPubSub {
...
 
-  async subscribe(pubsubName: string, topic: string, cb: TypeDaprPubSubCallback, route = "") {
+  async subscribe(pubsubName: string, topic: string, cb: TypeDaprPubSubCallback, route = "", metadata?: KeyValueType) {
     if (!route) {
       route = `route-${pubsubName}-${topic}`;
     }
 
     // Register the handler
-    await this.server.getServerImpl().registerPubSubSubscriptionRoute(pubsubName, topic, route);
+    await this.server.getServerImpl().registerPubSubSubscriptionRoute(pubsubName, topic, route, metadata ?? {});
```
Comment (XavierGeerinck) [inline src/implementation/Server/HTTPServer/pubsub.ts:37]: Shouldn't we just pass metadata and make it default in the `registerPubSubSubscriptionRoute`?

File: src/utils/Client.util.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import * as grpc from "@grpc/grpc-js";
+import { KeyValueType } from "../types/KeyValue.type";
+
+/**
+ * Converts a KeyValueType to a grpc.Metadata object.
+ * @param metadata key value pair of metadata
+ * @returns grpc.Metadata object
+ */
+export function getGRPCMetadata(metadata: KeyValueType): grpc.Metadata {
```
Comment (XavierGeerinck) [inline src/utils/Client.util.ts:22]: Could we rename to `generateGRPCMetadata`? sounds more like a generate instead of a getter

Also add default metadata and param check?

File: src/utils/Client.util.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * 
+ * Example, if the metadata is { "key1": "value1", "key2": "value2" }, the query parameter will be:
+ * "metadata.key1=value1&metadata.key2=value2"
+ * 
+ * Note, the returned value does not contain the "?" prefix.
+ * 
+ * @param metadata key value pair of metadata
+ * @returns HTTP query parameter string
+ */
+export function getHTTPMetadataQueryParam(metadata: KeyValueType): string {
```
Comment (XavierGeerinck) [inline src/utils/Client.util.ts:43]: Could we rename to `generateHTTPMetadataQueryParam`? sounds more like a generate instead of a getter.

Also add default metadata and param check?

File: test/unit/utils/Client.util.test.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {getGRPCMetadata, getHTTPMetadataQueryParam} from "../../../src/utils/Client.util"
+
+describe('Client.util', () => {
```
Comment (XavierGeerinck) [inline test/unit/utils/Client.util.test.ts:16]: * Also add when metadata is null or undefined check
* Also add hashbang == '#' (e.g. http://www.facebook.com/example.profile#!/pages/Another-Page/123456789012345)

**Review by XavierGeerinck — CHANGES_REQUESTED** [review]

File: src/implementation/Server/GRPCServer/pubsub.ts (hunk 2 above)
Comment (shubham1172) [reply src/implementation/Server/GRPCServer/pubsub.ts:32]: Same as above, let me know if you want me to change it.

File: src/implementation/Client/HTTPClient/pubsub.ts (hunk 1 above)
Comment (XavierGeerinck) [reply src/implementation/Client/HTTPClient/pubsub.ts:31]: But it defaults through the method signature that adds a ? No?

File: src/utils/Client.util.ts (hunk 4 above)
Comment (shubham1172) [reply src/utils/Client.util.ts:22]: It's more of a "converter" than "generator". Shall we call it `GRPCMetadataFromKeyValueType` and `HTTPMetadataQueryParamFromKeyValuePairType`? https://stackoverflow.com/questions/15962550/function-name-convention-for-convert-foo-to-bar

This might be too verbose though. Thoughts?

File: src/implementation/Client/HTTPClient/pubsub.ts (hunk 1 above)
Comment (shubham1172) [reply src/implementation/Client/HTTPClient/pubsub.ts:31]: I see where you are coming from, sure! Let me update it.

File: src/utils/Client.util.ts (hunk 4 above)
Comment (XavierGeerinck) [reply src/utils/Client.util.ts:22]: or `createGRPCMetadata`? in the end I think I might be fine with: `convertKeyValueToGRPCMetadata` verbosity, yet it's an issue, but I'd rather have it descriptive than confusing :D

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 7]
```diff
@@ -60,9 +60,9 @@ export default class GRPCServerImpl implements IAppCallbackServer {
         this.handlersInvoke[handlerKey] = cb;
     }
 
-    registerPubSubSubscriptionHandler(pubSubName: string, topicName: string, cb: TypeDaprInvokerCallback, metadata: KeyValueType): void {
+    registerPubSubSubscriptionHandler(pubSubName: string, topicName: string, cb: TypeDaprInvokerCallback, metadata?: KeyValueType): void {
         const handlerKey = this.createPubSubSubscriptionHandlerKey(pubSubName, topicName);
-        this.registrationsTopics[handlerKey] = { cb: cb, metadata: metadata };
+        this.registrationsTopics[handlerKey] = { cb: cb, metadata: metadata ?? {} };
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:65]: No need for ?? I think, let's add `metadata?: KeyValueType = {}` instead? this will use `{}` as default

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 8]
```diff
@@ -27,9 +27,9 @@ export default class HTTPServerImpl {
         this.pubSubSubscriptionRoutes = [];
     }
 
-    registerPubSubSubscriptionRoute(pubSubName: string, topicName: string, route: string, metadata: KeyValueType): void {
+    registerPubSubSubscriptionRoute(pubSubName: string, topicName: string, route: string, metadata?: KeyValueType): void {
```
Comment (XavierGeerinck) [inline src/implementation/Server/HTTPServer/HTTPServerImpl.ts:30]: Also use `metadata?: KeyValueType = {}` here

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 9]
```diff
@@ -27,9 +27,9 @@ export default class HTTPServerImpl {
...
         const httpMetadata: { [key: string]: string } = {};
-        for (const [key, value] of Object.entries(metadata)) {
+        for (const [key, value] of Object.entries(metadata ?? {})) {
```
Comment (XavierGeerinck) [inline src/implementation/Server/HTTPServer/HTTPServerImpl.ts:32]: Not needed as it should be set correctly in the method signature and validated straight under it

File: src/utils/Client.util.ts [hunk 10]
```diff
@@ -40,9 +40,9 @@ export function getGRPCMetadata(metadata: KeyValueType): grpc.Metadata {
  * @param metadata key value pair of metadata
  * @returns HTTP query parameter string
  */
-export function getHTTPMetadataQueryParam(metadata: KeyValueType): string {
+export function getHTTPMetadataQueryParam(metadata?: KeyValueType): string {
     let queryParam = "";
-    for (const [key, value] of Object.entries(metadata)) {
+    for (const [key, value] of Object.entries(metadata ?? {})) {
```
Comment (XavierGeerinck) [inline src/utils/Client.util.ts:45]: Also use `metadata?: KeyValueType = {}` here

**Review by XavierGeerinck — CHANGES_REQUESTED** [review]
Some more requests, also, do we have a test for if metadata is null or undefined? (Javascript tends to be annoying with it)

File: src/utils/Client.util.ts (hunk 4 above)
Comment (shubham1172) [reply src/utils/Client.util.ts]: Thanks @XavierGeerinck, went with `createXXX` then.

**Comment by shubham1172** [comment]
> Some more requests, also, do we have a test for if metadata is null or undefined? (Javascript tends to be annoying with it)

@XavierGeerinck, TS won't allow you to pass null. Tests for undefined and empty are added.

**Review by XavierGeerinck — APPROVED** [review]

**Comment by chris-iliopoulos** [comment]
@XavierGeerinck Sorry if this is not the place to ask, but any plans on when will this fix be released? Thnx in advance!

**Comment by XavierGeerinck** [comment]
Hi there! No worries for asking! We have currently merged it and are working towards a release this week (We have a sync call tonight so release wil most likely followl).

**Comment by chris-iliopoulos** [comment]
Great! Thank you again ... looking forward to this fix!
